src/cargo/ops/cargo_install.rs

```diff
@@ -182,16 +197,48 @@ fn install_one(
             }
         }
         src.update()?;
-        select_pkg(src, krate, vers, config, false, &mut |path| {
+        select_pkg(&mut src, krate, None, config, false, &mut |path| {
             path.read_packages()
         })?
     } else {
+        let mut source = map.load(source_id, &HashSet::new())?;
+        let vers = if let Some(vers) = vers {
+            Some(parse_semver_req(vers)?)
+        } else if source.source_id().is_registry() {
+            Some(VersionReq::any())
+        } else {
+            None
+        };
+        if krate.is_some() && !no_track {
```

Comment:
I feel like this section could use some comments explaining what it is doing.  Particularly, the one important part (passing `false` to `select_pkg`) needs to be highlighted as being the important part, and *why* it is doing that.